sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -323,6 +324,19 @@ trait RuntimeReplaceable extends UnaryExpression with Unevaluable {
   // two `RuntimeReplaceable` are considered to be semantically equal if their "child" expressions
   // are semantically equal.
   override lazy val canonicalized: Expression = child.canonicalized
+
+  /**
+   * Only used to generate SQL representation of this expression.
+   *
+   * Implementations should override this with original parameters
+   */
+  def exprsReplaced: Seq[Expression]
+
+  override def sql: String = mkString(exprsReplaced.map(_.sql))
+
+  def mkString(childrenString: Seq[String]): String = {
+    prettyName + childrenString.mkString("(", ", ", ")")
+  }
```

Comment:
Why do we need two extra interface methods just to make the column name pretty? Seems the output isn't even completely pretty.

Reply:
It's not about pretty or not. It's about consistency. If we call a function with its alias name, then it will not use `PrettyAttribute` and the string representation is different from if we call the function with the original name.

Reply:
The consistency is good but the readability here is the problem. See the implementation side:

```scala
  override def sql: String = s"$prettyName(${field.sql} FROM ${source.sql})"
  override def prettyName: String = "extract"
```
vs
```scala
  override def exprsReplaced: Seq[Expression] = Seq(field, source)

  override def mkString(childrenString: Seq[String]): String = {
    prettyName + childrenString.mkString("(", " FROM ", ")")
  }
```

Previously, it was fine to only know `sql` and `prettyName` but now you should know what `exprsReplaced` and `mkString` are. From the naming, it's confusing what that means too.

Reply:
I would just document that `sql` and `prettyName` should be properly overridden at `RuntimeReplaceable` just like we have the restriction about the `child`.

Reply:
The caller side decides if he wants pretty string or not, by calling `Expression.sql` or `toPrettySQL(Expression)`. So simply overridden `sql` doesn't work, unless we add a boolean parameter.